### PR TITLE
GROOVY-9803: translate generics of method reference using SAM parameters

### PR DESCRIPTION
**Pull request by eric-milles**
Since this change resolves additional generics, the test case for GROOVY-9762 fails when comparing `List<? super Integer>` to `List<Integer>`.  Wherever `Collections::singletonList` is matched up with `Function<? super T, ? extends U>` is getting `List<? super Integer>`.  Java just shows `List<Integer>` for `Optional.of(123).map(Collections::singletonList)`.

https://issues.apache.org/jira/browse/GROOVY-9803

### CODE REVIEW

**Comment by eric-milles** [comment]
I don't know if the change in `applyGenericsContextToParameterClass` is the right thing to do.  I have no Java specification section to quote for how return type generics should be handled.  Java gives `Set<Integer>` for `Optional.of(123).map(Collections::singleton).get()`.  I read over [this](https://docs.oracle.com/javase/tutorial/extra/generics/morefun.html) but didn't figure any specific guidance for how `map` should work in this case.

**Note:** `applyGenericsContextToParameterClass` is only used by `StaticTypeCheckingVisitor`.  I would have moved it there if not for binary backwards compatibility.

File: src/main/java/org/codehaus/groovy/transform/stc/StaticTypeCheckingVisitor.java [hunk 1]
```diff
@@ -5272,6 +5275,46 @@ private static void extractGenericsConnectionsForSuperClassAndInterfaces(final M
...
+            if (types.length == paramTypes.length) {
+                for (int i = 0, n = types.length; i < n; i += 1) {
+                    if (!types[i].isGenericsPlaceHolder() && !isAssignableTo(types[i], paramTypes[i])) {
+                        return false;
+                    }
+                }
+                return true;
+            }
+            return false;
+        }).findFirst().orElse(null); // TODO: order matches by param distance
```
Comment (danielsun1106) [inline src/main/java/org/codehaus/groovy/transform/stc/StaticTypeCheckingVisitor.java:5296]: Are you going to tweak the PR further?

Comment (eric-milles) [reply src/main/java/org/codehaus/groovy/transform/stc/StaticTypeCheckingVisitor.java:5296]: I have no further planned changes.  It should be workable as it is, unless there is another test case that can demonstrate an issue.

Comment (danielsun1106) [reply src/main/java/org/codehaus/groovy/transform/stc/StaticTypeCheckingVisitor.java:5296]: Gotcha

**Comment by danielsun1106** [comment]
Merged. Thanks!

**Comment by chrylis** [comment]
I'm not sure from reading the code change, but does this handle arbitrarily nested generics? The actual real-world type that the compiler is having trouble with is (combining Vavr, Spring MVC, and some of my own record types)

```
<E, A, R> Either<JsonApiErrorResponse<E>, HeadersAndBody<JsonApiDataItemResponse<A, R>>>
```

where invoking `Either#map` expects a `Function<? super HeadersAndBody<JsonApiDataItemResponse<A, R>>` (this part works) but the closure (or Parrot lambda) seems not to correctly preserve the nested type parameters into the closure.

**Comment by eric-milles** [comment]
@chrylis You can try out your specific case using the CI build of groovy or the latest snapshot build of groovy-eclipse.

**Pull request merged**